### ISSUE TITLE
SSR issues (4.3.0)

### ISSUE THREAD

**Issue by slorber** [rpearce/image-zoom#245]
Hi,

I have SSR issues in a Gatsby (theme novela) due to this lib. 

I saw the SSR fixes, tried versions like 4.0 or 4.3 and it didn't improve my issue.

You can see it live in production at https://sebastienlorber.com/atomic-css-in-js
Repro is my own Gatsby site: https://github.com/slorber/sebastienlorber.com

Before hydration, the page looks like:

![image](https://user-images.githubusercontent.com/749374/81272707-1efd7f80-904e-11ea-9792-3bcb06c90b58.png)

After hydration, the page looks like:

![image](https://user-images.githubusercontent.com/749374/81272750-2f155f00-904e-11ea-8b7e-770ce8483626.png)

I know there is no image in the screenshot, but it's definitively the images that are the cause of this problem.

My Gatsby theme uses this lib here: https://github.com/narative/gatsby-theme-novela/blob/master/%40narative/gatsby-theme-novela/src/components/Image/Image.Zoom.tsx

If I tweak the source code, and make "Controlled.js" return "children", my markup issues get fixed.

Not sure exactly what's the problem, but that seems related to https://joshwcomeau.com/react/the-perils-of-rehydration

**Comment by rpearce** [comment]
@slorber You're in luck! I've got some time off work coming up after tomorrow, so I'll have time to put towards updating things here.

If you really want to help me resolve this issue ASAP, can you 🙏  create an example repository that only exposes this issue? That'll help me narrow in and resolve it so much faster

**Comment by slorber** [comment]
Hi,

Will try to get a better repro, but for now what I have is this minimal MDX on my site:


```
---
title: Atomic CSS-in-JS
slug: atomic-css-in-js
author: Sébastien Lorber
date: 2020-04-27
hero: ./images/hero.png
excerpt: ...or how to scale your CSS-in-JS.
secret: false
---


![](./images/rnw-rules-order.png)

```jsx
import {t} from 'react-native-tailwindcss';

<View style={[t.absolute, t.inset0, t.p4, t.bgBlue500]} />
```

**Note**: This predictable behavior is only possible thanks to using the most strict atomic CSS.


```

It seems I have to add both the image zoom + the code to reproduce this issue. Will try to see how to isolate that more but it could take time.

**Comment by rpearce** [comment]
FYI I'm almost done completely rewriting this library into:
1. a vanilla JS package that is ~3kb (minified & gzipped)
1. a react hook that uses 👆  and is maybe 100-150 bytes 

This and all the other issues should be closed out by it. I'm _almost_ there (working out some API stuff).

Thanks for your patience

**Comment by rpearce** [comment]
There is an alpha release out (no documentation nor tests yet): [5.0.0-alpha.1](https://www.npmjs.com/package/react-medium-image-zoom/v/5.0.0-alpha.1)

This has condensed the "react" nature of this to nothing more than a react hook wrapping https://www.npmjs.com/package/@rpearce/image-zoom/v/0.1.0-alpha.2.

For gatsby, it looks like their `gatsby-image` might not accept `ref` (let me know if it does), which sucks, but good news: you can do this:
```js
const { ref } = useImageZoom({ /* image zoom options here */ })

return (
  <div ref={ref}>
    <Image ... />
  </div>
```
However, it is not guaranteed to have all the refined parts that using it directly on an `img` tag would.

I'm considering doing a check to see if the current target is NOT an `img`, but the child IS an `img`, then just attach it to that -- maybe through a `type` flag with the value `'find-first-img'` or something, but I haven't thought it out fully.

**Comment by 340rahul** [comment]
This looks great! The github repo link on image-zoom's npm page gives a 404. Can you please link the repo here?

**Comment by rpearce** [comment]
The homepage should now work for it; some life stuff came up, and I have a lot of build work to do on this and a couple of issues to resolve before I can get everything in sync.

**Comment by slorber** [comment]
thanks @rpearce 

I'm closing because actually updating my gatsby / novela did fix the SSR issue I had, so it was probably not related to this lib finally...

(fixed while still on 4.3.1)
